src/btf.cpp

```diff
@@ -38,151 +39,116 @@ static __u32 type_cnt(const struct btf *btf)
+                        va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
 
-  size = st.st_size;
+BTF::BTF(const std::set<std::string> &modules) : state(NODATA)
+{
+  if (bt_debug != DebugLevel::kNone)
+    libbpf_set_print(libbpf_print);
 
-  data = (unsigned char *) malloc(size);
-  if (!data)
+  // Try to get BTF file from BPFTRACE_BTF env
+  char *path = std::getenv("BPFTRACE_BTF");
+  if (path)
   {
-    fclose(f);
-    return nullptr;
+    btf_objects.emplace_back(btf__parse_raw(path), 0, "");
+    vmlinux_btf = btf_objects.back().btf;
   }
+  else
+    load_kernel_btfs(modules);
 
-  ssize_t ret = fread(data, 1, st.st_size, f);
-
-  if (ret != st.st_size)
+  if (btf_objects.empty())
   {
-    free(data);
-    fclose(f);
-    return nullptr;
+    if (bt_debug != DebugLevel::kNone)
+      LOG(ERROR) << "BTF: failed to find BTF data";
+    return;
   }
 
-  fclose(f);
-
-  *sizep = size;
-  return data;
-}
-
-static struct btf* btf_raw(char *file)
-{
-  unsigned char *data;
-  ssize_t size;
-  struct btf *btf;
-
-  data = get_data(file, &size);
-  if (!data)
-  {
-    LOG(ERROR) << "BTF: failed to read data from: " << file;
-    return nullptr;
-  }
+  vmlinux_btf_size = (__s32)type_cnt(vmlinux_btf);
 
-  btf = btf__new(data, (__u32) size);
-  free(data);
-  return btf;
+  state = OK;
 }
 
-static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+BTF::~BTF()
 {
-  fprintf(stderr, "BTF: (%d) ", level);
-  return vfprintf(stderr, msg, ap);
+  for (auto &btf_obj : btf_objects)
+    btf__free(btf_obj.btf);
 }
 
-static struct btf *btf_open(const struct vmlinux_location *locs)
+void BTF::load_kernel_btfs(const std::set<std::string> &modules)
 {
-  struct utsname buf;
-
-  uname(&buf);
-
-  for (int i = 0; locs[i].path; i++)
+  // Note that we cannot parse BTFs from /sys/kernel/btf/ as we need BTF object
+  // IDs, so the only way is to iterate through all loaded BTF objects
+  __u32 id = 0;
+  while (true)
   {
-    char path[PATH_MAX + 1];
+    int err = bpf_btf_get_next_id(id, &id);
+    if (err)
+    {
+      if (errno != ENOENT && bt_debug != DebugLevel::kNone)
+        LOG(WARNING) << "BTF: failed to iterate modules BTF objects";
+      break;
+    }
 
-    snprintf(path, PATH_MAX, locs[i].path, buf.release);
-    if (access(path, R_OK))
+    // Get BTF object FD
+    int fd = bpf_btf_get_fd_by_id(id);
+    if (fd < 0)
+    {
+      if (bt_debug != DebugLevel::kNone)
+        LOG(WARNING) << "BTF: failed to get FD for object with id " << id;
       continue;
+    }
 
-    struct btf *btf;
-
-    if (locs[i].raw)
-      btf = btf_raw(path);
-    else
-      btf = btf__parse_elf(path, nullptr);
-
-    int err = libbpf_get_error(btf);
+    // Get BTF object info - needed to determine if this is a kernel module BTF
+    char name[64];
+    struct bpf_btf_info info;
+    __u32 info_len = sizeof(info);
+    memset(&info, 0, info_len);
```

Comment:
Prefer C++ default init: 
```suggestion
    struct bpf_btf_info info{};
```